### PR TITLE
fix: make_uind handles empty ECORR bases and computes shape in numpy

### PR DESCRIPTION
**Pull request by vhaasteren**
## Summary

- Fix `make_uind()` crashing when the ECORR quantization matrix has zero epochs (`U.shape[1] == 0`), e.g. when an ECORR selection matches no TOAs. Return an empty `(0, 1)` integer index table instead of calling `jnp.max` on an empty array.
- Compute the max epoch size with NumPy (`int(np.max(np.sum(U, axis=0)))`) rather than JAX, since `make_uind` builds setup-time index tables whose shapes must be concrete Python integers, not traced JAX scalars.
- Add direct unit tests for variable epoch sizes (zero-padded 1-based TOA rows) and the empty-basis edge case.

## Problem

`make_uind(U)` builds the indexed ECORR lookup table used by the Schur-complement noise-matrix kernels (`NoiseMatrixSM_*` → `SM_*_indexed`). On `main`, it sizes the output with:

```python
Uind = np.zeros((U.shape[1], jnp.max(jnp.sum(U, axis=0)) + 1), 'i')
```

Two issues:

1. **Empty basis:** If an ECORR group selects no TOAs, `U` has shape `(ntoa, 0)`. `jnp.max` over the empty per-epoch counts raises and model setup fails.
2. **JAX tracing at setup time:** Using `jnp.max` for array allocation can produce traced scalars where concrete integer shapes are required. This is setup-time bookkeeping, not a quantity that should be JAX-traced.

## Solution

- Early-return `np.zeros((0, 1), 'i')` when `U.shape[1] == 0`.
- Coerce input with `np.asarray(U)` and compute `maxcount` with NumPy before allocating `Uind`.

Behavior for non-empty bases is unchanged: each row holds 1-based TOA indices for that epoch, zero-padded to the widest epoch.

## Test plan

- [x] `pytest tests/test_matrix.py::TestMakeUind -v`
- [x] `pytest tests/test_matrix.py -v` (full matrix test module)
- [x] Confirm a pulsar/model setup path that includes an ECORR selection matching zero TOAs no longer crashes during `make_uind` / noise-matrix construction

### CODE REVIEW

**Comment by danielreardon** [comment]
Thanks @vhaasteren for this. We encountered the same bug and implemented a fix that is a subset of your PR: https://github.com/danielreardon/discovery/commit/041a4a6e6942b96e03cdd25351e261841c269b73

So I'd like to merge this version and @meyers-academic can consider this an endorsement/review! :)

**Comment by meyers-academic** [comment]
This looks good to me so far. Shall I wait for the empty TOA test before approving/merging?

**Review by meyers-academic — APPROVED** [review]
I accept @danielreardon 's endorsement :).
